fvm/src/gas/mod.rs

```diff
@@ -194,10 +171,18 @@ pub struct GasTracker {
     trace: Option<RefCell<Vec<GasCharge>>>,
 }
 
+const UNLIMITED_GAS: Gas = Gas::from_milligas(i64::MAX as u64);
+
 impl GasTracker {
     /// Gas limit and gas used are provided in protocol units (i.e. full units).
     /// They are converted to milligas for internal canonical accounting.
-    pub fn new(gas_limit: Gas, gas_used: Gas, enable_tracing: bool) -> Self {
+    ///
+    /// - If the gas limit exceeds `i64::MAX` milligas, it's rounded down to `i64::MAX` milligas and
+    ///   treated as "unlimited".
+    /// - If the gas used exceeds the gas limit, it's capped at the gas limit.
+    pub fn new(mut gas_limit: Gas, mut gas_used: Gas, enable_tracing: bool) -> Self {
+        gas_limit = gas_limit.min(UNLIMITED_GAS);
+        gas_used = gas_used.min(gas_limit);
```

Comment:
I changed this to be consistent. That way, if we have "too much" gas, we just don't burn gas. Otherwise, there were a lot of edge-cases around the maximums that were hard to reason about.